File: arthur/backend/media/ArthurColor.java
package arthur.backend.media;

import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.awt.*;
import java.awt.image.*;
import java.awt.geom.*;
import java.awt.font.*;
import java.util.*;

import arthur.backend.builtins.java.*;

/**
 * Java implementation of arthur color!
 */
public class ArthurColor extends ArthurMedia {

  public static final String COLOR = "color";
  public ArthurNumber r, g, b, a;

  public ArthurColor() {
    this(new ArthurNumber(0), new ArthurNumber(0), new ArthurNumber(0), new ArthurNumber(0));
  }

  public ArthurColor(double r, double g, double b, double a) {
    this(new ArthurNumber(r), new ArthurNumber(g), new ArthurNumber(b), new ArthurNumber(a));
  }

  public ArthurColor(ArthurNumber r, ArthurNumber g, ArthurNumber b, ArthurNumber a) {
    this.type = COLOR;
    this.r=r;
    this.g=g;
    this.b=b;
    this.a=a;
  }

  public ArthurColor flip() {
    return new ArthurColor(this.b, this.g, this.r, this.a);
  }

  public ArthurColor opposite() {
    return new ArthurColor(255 - this.r.val, 255 - this.g.val, 255 - this.b.val, this.a.val);
  }

  public ArthurNumber valDiff(ArthurColor other) {
    double rd = Math.abs(this.r.val - other.r.val);
    double gd = Math.abs(this.g.val - other.g.val);
    double bd = Math.abs(this.b.val - other.b.val);
    double ad = Math.abs(this.a.val - other.a.val);
    return new ArthurNumber(rd + gd + bd + ad);
  }

  public ArthurColor add(ArthurMedia two) {
    if (two.type.equals(COLOR)) {
      return JavaColorMath.add(this, (ArthurColor)two);
    } else if (two.type.equals(ArthurImage.IMAGE)) {
      return JavaColorMath.add(this, (ArthurImage)two);
    }
    else if (two.type.equals(ArthurSound.SOUND)) {
      return JavaColorMath.add(this, (ArthurSound)two);
    }
    else if (two.type.equals(ArthurVideo.VIDEO)) {
      return JavaColorMath.add(this, (ArthurVideo)two);
    }
    else if (two.type.equals(ArthurString.STRING)) {
      return JavaColorMath.add(this, (ArthurString)two);
    }
    else if (two.type.equals(ArthurNumber.NUMBER)) {
      return JavaColorMath.add(this, (ArthurNumber)two);
    }
    else {
      return this;
    }
  }

  public ArthurColor minus(ArthurMedia two) {
    if (two.type.equals(COLOR)) {
      return JavaColorMath.minus(this, (ArthurColor)two);
    } else if (two.type.equals(ArthurImage.IMAGE)) {
      return JavaColorMath.minus(this, (ArthurImage)two);
    }
    else if (two.type.equals(ArthurSound.SOUND)) {
      return JavaColorMath.minus(this, (ArthurSound)two);
    }
    else if (two.type.equals(ArthurVideo.VIDEO)) {
      return JavaColorMath.minus(this, (ArthurVideo)two);
    }
    else if (two.type.equals(ArthurString.STRING)) {
      return JavaColorMath.minus(this, (ArthurString)two);
    }
    else if (two.type.equals(ArthurNumber.NUMBER)) {
      return JavaColorMath.minus(this, (ArthurNumber)two);
    }
    return this;
  }

  public ArthurColor multiply(ArthurMedia two) {
    if (two.type.equals(COLOR)) {
      return JavaColorMath.multiply(this, (ArthurColor) two);
    } else if (two.type.equals(ArthurNumber.NUMBER)) {
      return JavaColorMath.multiply(this, (ArthurNumber) two);
    } else if (two.type.equals(ArthurImage.IMAGE)) {
      ArthurImage img = (ArthurImage) two;
      return JavaColorMath.multiply(this, img.toColor());
    } else if (two.type.equals(ArthurString.STRING)) {
      ArthurString str = (ArthurString) two;
      return JavaColorMath.multiply(this, str.toColor());
    }
    return this;
  }

  public ArthurColor divide(ArthurMedia two) {
    if (two.type.equals(COLOR)) {
      return JavaColorMath.divide(this, (ArthurColor) two);
    } else if (two.type.equals(ArthurNumber.NUMBER)) {
      return JavaColorMath.divide(this, (ArthurNumber) two);
    } else if (two.type.equals(ArthurImage.IMAGE)) {
      ArthurImage img = (ArthurImage) two;
      return JavaColorMath.divide(this, img.toColor());
    } else if (two.type.equals(ArthurString.STRING)) {
      ArthurString str = (ArthurString) two;
      return JavaColorMath.divide(this, str.toColor());
    }
    return this;
  }

  public ArthurMedia castTo(ArthurString mediaType) {
    return castTo(mediaType.str);
  }

  public ArthurMedia castTo(String mediaType) {
    if (mediaType.equals("string")) {
      return this.toArtString();
    } else if (mediaType.equals("number")) {
      return this.toNumber();
    } else if (mediaType.equals("Image")) {
      return this.toImage();
    }

    return this;
  }

  public ArthurNumber toNumber() {
    double val = 0;
    val += this.g.val;
    val += 1000 * this.b.val;
    val += 1000000 * this.r.val;
    return new ArthurNumber(val);
  }

  public ArthurImage toImage() {
    String filename = "color-" + (new Random()).nextInt() + ".jpg";

    int w = 300;
    int h = 300;

    BufferedImage image = new BufferedImage(w, h, BufferedImage.TYPE_INT_RGB);

    Graphics2D g = image.createGraphics();
    Color c = new Color(this.r.val.floatValue(), this.g.val.floatValue(), this.b.val.floatValue(), this.a.val.floatValue());
    g.setColor(c);
    g.fillRect(0, 0, w, h);
    g.dispose();

    return new ArthurImage(image, filename);
  }

  public ArthurVideo toVideo() {
    return JavaVideoMath.add(new ArthurVideo("ZERO.mp4"), this, "color-" + System.currentTimeMillis() + ".mp4");
  }

  public ArthurString toArtString() {
    return closestString();
  }

  public ArthurString closestString() {
    ArthurNumber bestDiff = new ArthurNumber(Double.POSITIVE_INFINITY);
    ArthurNumber diff;
    ArthurColor current;
    ArthurString best = new ArthurString("color");

    for (String c : JavaBuiltins.colors()) {
      current = JavaBuiltins.colorMap().get(c);
      diff = this.valDiff(current);

      if (diff.val < bestDiff.val) {
        bestDiff = diff;
        best = new ArthurString(c);
      }
    }

    return best;
  }

  public ArthurString furthestString() {
    ArthurNumber bestDiff = new ArthurNumber(Double.NEGATIVE_INFINITY);
    ArthurNumber diff;
    ArthurColor current;
    ArthurString best = new ArthurString("color");

    for (String c : JavaBuiltins.colors()) {
      current = JavaBuiltins.colorMap().get(c);
      diff = this.valDiff(current);

      if (diff.val > bestDiff.val) {
        bestDiff = diff;
        best = new ArthurString(c);
      }
    }

    return best;
  }

  public boolean arthurEquals(ArthurMedia two) {
    if (two.type.equals(COLOR)) {
      ArthurColor t = (ArthurColor) two;
      if (this.r.arthurEquals(t.r) &&
          this.g.arthurEquals(t.g) &&
          this.b.arthurEquals(t.b) &&
          this.a.arthurEquals(t.a))
          return true;

      return false;
    }

    return false;
  }

  public String toString() {
    return "<<" + this.r + ", " + this.g + ", " + this.b + ">>";
  }

  public String jsLiteral() {
    String js = "new ArthurColor(";
    js += "'" + json() + "'";
    js += ")";
    return js;
  }

  public String json() {
    String json = "{'r': " + r.val + ", ";
    json += "'g': " + g.val + ", ";
    json += "'b': " + b.val + ", ";
    if (this.frame != null) {
      json += "'frame': " + this.frame.json() + ", ";
    }
    if (this.delay != null) {
      json += "'delay': " + this.delay.val + ", ";
    }
    json += "'a': " + a.val + "}";
    json = json.replace("'", "\"");
    return json;
  }

  /* saves state of current media to filename */
  public void writeToFile(String filename) {
    String json = json();
    try {
      PrintWriter out = new PrintWriter(filename);
      out.println(json);
      out.close();
    } catch(FileNotFoundException e) {
      e.printStackTrace();
    }
  }

}

File: samples/sample15trans/js/arthurbackup/arthur-color.js

var types = require('./types');
//var builtins = require('./builtins');
var ArthurNumber = require('./arthur-number');
var ArthurMedia = require('./arthur-media');
var ArthurFrame = require('./arthur-frame');

var canvas = document.getElementById('canvas');
var context = canvas.getContext('2d');

module.exports = ArthurColor;

console.log(module.exports);

ArthurColor.prototype.__proto__ = ArthurMedia.prototype;

/* assumes either all arthurnumbers, or all js numbers or one string*/
function ArthurColor(r, g, b, a, frame, delay) {
  this.type = types.COLOR;

  if (frame) {
    this.frame = new ArthurFrame(frame);
  }
  if (delay) {
    this.delay = new ArthurNumber(delay);
  }

  if (r instanceof ArthurNumber) {
    this.r = r;
    this.g = g;
    this.b = b;
    this.a = a;
  } else if (r instanceof ArthurColor) {
    this.r = r.r;
    this.g = r.g;
    this.b = r.b;
    this.a = r.a;
  } else if (typeof r == 'object') {
    this.r = new ArthurNumber(r.r);
    this.g = new ArthurNumber(r.g);
    this.b = new ArthurNumber(r.b);
    this.a = new ArthurNumber(r.a);
    this.frame = r.frame;
    if (r.delay) {
      this.delay = new ArthurNumber(r.delay);
    }
  } else if (typeof r == 'string') {
    var ob = JSON.parse(r);
    this.r = new ArthurNumber(ob.r);
    this.g = new ArthurNumber(ob.g);
    this.b = new ArthurNumber(ob.b);
    this.a = new ArthurNumber(ob.a);
    this.frame = ob.frame;
    if (ob.delay) {
      this.delay = new ArthurNumber(ob.delay);
    }
  } else {
    this.r = new ArthurNumber(r);
    this.g = new ArthurNumber(g);
    this.b = new ArthurNumber(b);
    this.a = new ArthurNumber(a);
  }
  this.medfile = null;

  if (!this.a) {
    this.a = new ArthurNumber(1.0);
  }
}

ArthurColor.prototype.rgba = function() {
  var rgb = 'rgba(' + this.r.int() + ', ';
  rgb += this.g.int() + ', ';
  rgb += this.b.int() + ', ';
  rgb += this.a.val + ')';
  return rgb;
}

ArthurColor.prototype.draw = function() {
  context.fillStyle = this.rgba();

  if (!this.frame) {
    this.frame = new ArthurFrame({x: 0, y: 0, w: canvas.width, h: canvas.height});
  }

  context.fillRect(this.frame.x.int(), this.frame.y.int(), this.frame.w.int(), this.frame.h.int());
}

ArthurColor.prototype.add = function(color) {
  if (color.type == types.COLOR) {
    var r = Math.min(this.r.val + color.r.val, 255);
    var g = Math.min(this.g.val + color.g.val, 255);
    var b = Math.min(this.b.val + color.b.val, 255);
    var a = this.a.val;

    var ob = {r: r, g: g, b: b, a: a, frame: this.frame, delay: this.delay};
    return new ArthurColor(ob);
  } else if (color.type == types.NUMBER) {
    var abs = Math.abs(color.val);
    var r = Math.max(this.r.val + abs, 255);
    var g = Math.max(this.g.val + abs, 255);
    var b = Math.max(this.b.val + abs, 255);
    var a = this.a.val;

    var ob = {r: r, g: g, b: b, a: a, frame: this.frame, delay: this.delay};
    return new ArthurColor(ob);
  } else if (color.type == types.STRING) {
    var c = color.castTo('color');
    return this.add(c);
  }

  return this;
}

ArthurColor.prototype.minus = function(color) {
  if (color.type == types.COLOR) {
    var r = Math.max(this.r.val - color.r.val, 0);
    var g = Math.max(this.g.val - color.g.val, 0);
    var b = Math.max(this.b.val - color.b.val, 0);
    var a = this.a.val;

    var ob = {r: r, g: g, b: b, a: a, frame: this.frame, delay: this.delay};
    return new ArthurColor(ob);
  } else if (color.type == types.NUMBER) {
    var abs = Math.abs(color.val);
    var r = Math.min(this.r.val - abs, 0);
    var g = Math.min(this.g.val - abs, 0);
    var b = Math.min(this.b.val - abs, 0);
    var a = this.a.val;

    var ob = {r: r, g: g, b: b, a: a, frame: this.frame, delay: this.delay};
    return new ArthurColor(ob);
  } else if (color.type == types.STRING) {
    var c = color.castTo('color');
    return this.minus(c);
  }

  return this;
}

ArthurColor.prototype.multiply = function(two) {
  if (two.type == types.COLOR) {
    var ob = {
      r: (this.r.val + two.r.val) / 2,
      g: (this.g.val+ two.g.val) / 2,
      b: (this.b.val+ two.b.val) / 2,
      a: Math.max(this.a.val, two.a.val),
      frame: this.frame,
      delay: this.delay
    };

    return new ArthurColor(ob);
  } else if (two.type == types.NUMBER) {
    var abs = Math.abs(two.val);
    var ob = {
      r: Math.max(this.r.val * abs, 255),
      g: Math.max(this.g.val * abs, 255),
      b: Math.max(this.b.val * abs, 255),
      a: Math.min(this.a.val * abs, 1.0),
      frame: this.frame,
      delay: this.delay
    };
    return new ArthurColor (ob);
  } else if (color.type == types.STRING) {
    var c = color.castTo('color');
    return this.multiply(c);
  }

  return this;
}

ArthurColor.prototype.divide = function(two) {
  if (two.type == types.COLOR) {
    var ob = {
      r: this.r.val / Math.max(two.r.val, 1),
      g: this.g.val / Math.max(two.g.val, 1),
      b: this.b.val / Math.max(two.b.val, 1),
      a: Math.max(this.a.val, two.a.val),
      frame: this.frame,
      delay: this.delay
    };
    return new ArthurColor(ob);
  } else if (two.type == types.NUMBER) {
    var abs = Math.abs(two.val);
    var ob = {
      r: this.r.val / Math.max(abs, 1),
      g: this.g.val / Math.max(abs, 1),
      b: this.b.val / Math.max(abs, 1),
      a: Math.min(this.a.val * abs, 1.0),
      frame: this.frame,
      delay: this.delay
    };
    return new ArthurColor (ob);
  } else if (color.type == types.STRING) {
    var c = color.castTo('color');
    return this.divide(c);
  }

  return this;
}

ArthurColor.prototype.valDiff = function(other) {
  if (!other.type || other.type != types.COLOR)
    return new ArthurNumber(Infinity);

  var rd = Math.abs(this.r.val - other.r.val);
  var gd = Math.abs(this.g.val - other.g.val);
  var bd = Math.abs(this.b.val - other.b.val);
  var ad = Math.abs(this.a.val - other.a.val);
  return new ArthurNumber(rd + gd + bd + ad);
}

ArthurColor.prototype.castTo = function(t) {
  if (t == 'string') {
    return this.closestString();
  }

  if (t == 'num') {
    var val = 0;
    val += this.g.val;
    val += 1000 * this.b.val;
    val += 1000000 * this.r.val;
    return new ArthurNumber(val);
  }

  return this;
}

ArthurColor.prototype.closestString = function() {
  var bestDiff = new ArthurNumber(Infinity);
  var diff;
  var currentColor, currentString;
  var bestColorString = new ArthurString("color");

  for (var i = 0; i < builtins.colorList.length; i++) {
    currentString = builtins.colorList[i];
    currentColor = builtins.colorMap[currentString];
    diff = this.valDiff(currentColor);

    if (diff.val < bestDiff.val) {
      bestDiff = diff;
      bestColorString = new ArthurString(currentString);
    }
  }

  return bestColorString;
}

ArthurColor.prototype.furthestString = function() {
  var bestDiff = new ArthurNumber(-1 * Infinity);
  var diff;
  var currentColor, currentString;
  var bestColorString = new ArthurString("color");

  for (var i = 0; i < builtins.colorList.length; i++) {
    currentString = builtins.colorList[i];
    currentColor = builtins.colorMap[currentString];
    diff = this.valDiff(currentColor);

    if (diff.val > bestDiff.val) {
      bestDiff = diff;
      bestColorString = new ArthurString(currentString);
    }
  }

  return bestColorString;
}
